File: bindings/dart/lib/proven.dart
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/// Proven - Dart FFI bindings to libproven (formally verified safety library).
///
/// All computation is performed in Idris 2 with dependent types and totality
/// checking. This package is a thin wrapper that marshals arguments across the
/// C ABI boundary via dart:ffi.
///
/// Requires libproven shared library (libproven.so / libproven.dylib /
/// proven.dll) to be available on the library load path.
///
/// Modules:
/// - [SafeMath] - Overflow-checked arithmetic
/// - [SafeString] - XSS prevention and string escaping
/// - [SafeFloat] - NaN/Infinity-safe floating-point operations
/// - [SafePath] - Directory traversal protection
/// - [SafeEmail] - Email validation
/// - [SafeNetwork] - IP address parsing and classification
/// - [SafeCrypto] - Constant-time comparison, secure random bytes
/// - [SafeHex] - Hexadecimal encoding and decoding
/// - [SafeUuid] - UUID v4 generation and validation (RFC 4122)
/// - [SafeJson] - JSON validation
/// - [SafeColor] - Color parsing and conversion
/// - [SafeAngle] - Angle unit conversion and normalization
/// - [SafeUnit] - Physical unit conversions
/// - [SafeCurrency] - Currency parsing (ISO 4217)
/// - [SafePhone] - Phone number parsing (E.164)
/// - [SafeDateTime] - Date/time parsing and calendar queries
/// - [SafeVersion] - Semantic version parsing (SemVer 2.0.0)
/// - [SafeUrl] - URL encoding and decoding
library proven;

export 'src/ffi.dart' show ProvenStatus;
export 'src/safe_angle.dart';
export 'src/safe_color.dart';
export 'src/safe_crypto.dart';
export 'src/safe_currency.dart';
export 'src/safe_datetime.dart';
export 'src/safe_email.dart';
export 'src/safe_float.dart';
export 'src/safe_hex.dart';
export 'src/safe_json.dart';
export 'src/safe_math.dart';
export 'src/safe_network.dart';
export 'src/safe_path.dart';
export 'src/safe_phone.dart';
export 'src/safe_string.dart';
export 'src/safe_unit.dart';
export 'src/safe_url.dart';
export 'src/safe_uuid.dart';
export 'src/safe_version.dart';

File: bindings/nim/src/proven.nim
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
#
# Proven -- FFI bindings to libproven (formally verified safety primitives).
#
# All computation is performed in the Idris2/Zig core via the C ABI exposed by
# libproven.  This Nim package is a thin wrapper that marshals data across
# the FFI boundary and returns idiomatic Nim types (Option for errors).
#
# Link with: -lproven

import proven/lib_proven
export lib_proven

import proven/safe_math
import proven/safe_string
import proven/safe_path
import proven/safe_email
import proven/safe_url
import proven/safe_network
import proven/safe_crypto
import proven/safe_uuid
import proven/safe_currency
import proven/safe_phone
import proven/safe_hex
import proven/safe_json
import proven/safe_datetime
import proven/safe_float
import proven/safe_version
import proven/safe_color
import proven/safe_angle
import proven/safe_unit

export safe_math
export safe_string
export safe_path
export safe_email
export safe_url
export safe_network
export safe_crypto
export safe_uuid
export safe_currency
export safe_phone
export safe_hex
export safe_json
export safe_datetime
export safe_float
export safe_version
export safe_color
export safe_angle
export safe_unit
